sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala

```diff
@@ -791,8 +799,26 @@ case class AlterTableRecoverPartitionsCommand(
       logDebug(s"Recovered ${parts.length} partitions ($done/$total so far)")
     }
   }
-}
 
+  // Drops the partitions that do not exist in the file system
+  private def dropPartitions(catalog: SessionCatalog, fs: FileSystem): Int = {
+    val dropPartSpecs = ThreadUtils.parmap(
+      catalog.listPartitions(tableName),
+      "AlterTableRecoverPartitionsCommand: non-existing partitions",
+      maxThreads = 8) { partition =>
+      partition.storage.locationUri.flatMap { uri =>
+        if (fs.exists(new Path(uri))) None else Some(partition.spec)
+      }
+    }.flatten
+    catalog.dropPartitions(
+      tableName,
+      dropPartSpecs,
+      ignoreIfNotExists = true,
+      purge = false,
+      retainData = true)
```

Comment:
Could you add a comment about the reason why we use `retainData=true`? I guess the reason is that `fs.exists(..)` is already `false` and we don't want addition file system calls. Did I understand correctly?

Reply:
> ... we don't want addition file system calls. Did I understand correctly?

Yep, if we set `retainData` to `true`, the `deleteData` flag will `false` at https://github.com/apache/hive/blob/release-3.1.3-rc0/standalone-metastore/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStore.java#L4360-L4378 . So, Hive MeteStore will not try to delete the partition folders.

Reply:
The same for the `In-Memory` catalog: https://github.com/apache/spark/blob/bfc023501379d28ae2db8708928f4e658ccaa07f/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/InMemoryCatalog.scala#L464-L473